packages/map/src/marker/index.jsx

```diff
@@ -13,37 +13,47 @@ const Marker = ({
   popupDefaultState,
   popupContent,
   popupOffset,
-  ...rest
 }) => {
   const nodeRef = useRef();
   const popupRef = useRef();
-  const map = useContext(MapContext);
-  useChildrenBounds([position]);
+  const [, forceRender] = useState();
 
-  useEffect(() => {
-    if (!map) return;
+  useChildrenBounds([position]);
 
+  useMapEffect((map) => {
+    nodeRef.current = document.createElement('div');
     const marker = new MapboxMarker(nodeRef.current).setLngLat(position).addTo(map);
 
     let popup;
     if (popupContent) {
+      popupRef.current = document.createElement('div');
       popup = new Popup({ offset: popupOffset }).setDOMContent(popupRef.current);
       marker.setPopup(popup);
       if (popupDefaultState) marker.togglePopup();
     }
 
+    forceRender({});
+
     return () => {
       if (popup) popup.remove();
       marker.remove();
+      nodeRef.current = null;
+      popupRef.current = null;
+      forceRender({});
     };
-  }, [map, children, popupContent, popupOffset]);
-
-  return (
-    <>
-      {popupContent && <div ref={popupRef}>{popupContent}</div>}
-      <div {...rest} ref={nodeRef}>{children}</div>
-    </>
-  );
+  }, [children, popupContent, popupOffset]);
+
+  let popupNode;
+  if (popupRef.current) {
+    popupNode = createPortal(popupContent, popupRef.current);
+  }
+
+  let markerNode;
+  if (nodeRef.current) {
+    markerNode = createPortal(children, nodeRef.current);
+  }
+
+  return <>{popupNode}{markerNode}</>;
```

Comment:
what was the problem with just rendering these div nodes with react?

Reply:
mapbox moves marker/popup nodes to body. because of that react throws error on unmount

Reply:
please add a comment